Adding SHA1 gadget

avras opened the pull request:
A second attempt at the SHA1 gadget using a simpler OR gate. See previous attempt at https://github.com/lurk-lab/bellpepper-gadgets/pull/6

avras commented:
I will make changes to this gadget to use the in-built `Boolean::or` once https://github.com/lurk-lab/bellpepper/pull/28 gets merged. We can hold this PR until then.

huitseeker commented:
Bellpepper 0.3.0 is released with the [or gadget](https://docs.rs/bellpepper/0.3.0/bellpepper/gadgets/boolean/enum.Boolean.html#method.or)!

avras commented:
Thanks @huitseeker for the heads up. It feels good to contribute to bellpepper. I will get the SHA1 PR done soon.

adr1anh commented on crates/sha1/src/sha1.rs:
```diff
@@ -0,0 +1,406 @@
+    Ok(cur.into_iter().flat_map(|e| e.into_bits_be()).collect())
+}
+
+fn get_sha1_iv() -> Vec<UInt32> {
+    IV.iter().map(|&v| UInt32::constant(v)).collect()
+}
```
The length of the result has a size known at compile time. Have you considered returning an array instead?

avras replied:
Thanks for pointing this out. I will try changing the output to an array

adr1anh commented on crates/sha1/src/sha1.rs:
```diff
@@ -0,0 +1,406 @@
+        .collect())
+}
+
+pub fn sha1<Scalar, CS>(mut cs: CS, input: &[Boolean]) -> Result<Vec<Boolean>, SynthesisError>
```
The function could return an array instead of a `Vec` since the size is known at compile-time.

avras replied:
I will try changing the output type here as well

adr1anh commented on crates/sha1/src/sha1.rs:
```diff
@@ -0,0 +1,406 @@
+            b.xor(cs.namespace(|| "b xor c"), &c)?
+                .xor(cs.namespace(|| "b xor c xor d"), &d)?
+        } else {
+            let a1 = and_uint32(cs.namespace(|| "1st and"), &b, &c)?;
+            let a2 = and_uint32(cs.namespace(|| "2nd and"), &b, &d)?;
+            let a3 = and_uint32(cs.namespace(|| "3rd and"), &c, &d)?;
+
+            let tmp = or_uint32(cs.namespace(|| "1st or"), &a1, &a2)?;
+            or_uint32(cs.namespace(|| "2nd or"), &tmp, &a3)?
```
The namespaces could describe the operations being performed, and a comment at the top of this block could state what the round computes.

avras replied:
I will revisit the SHA1 spec and give some descriptive labels to the namespaces. You can review and let me know if those are fine

adr1anh commented on crates/sha1/src/sha1.rs:
```diff
@@ -0,0 +1,406 @@
+        let f = if i < 20 {
+            // f = (b and c) or ((not b) and d)
+            UInt32::sha256_ch(cs.namespace(|| "ch"), &b, &c, &d)?
+        } else if !(40..60).contains(&i) {
```
A comment stating that this branch is for rounds `20 <= t <= 39` and `60 <= t <= 79` would make it easier to match the spec.

avras replied:
I will add this comment

adr1anh commented on crates/sha1/src/sha1.rs:
```diff
@@ -0,0 +1,406 @@
+            // b xor c xor d
+            b.xor(cs.namespace(|| "b xor c"), &c)?
+                .xor(cs.namespace(|| "b xor c xor d"), &d)?
+        } else {
```
Suggestion for a comment saying this branch is for rounds `40 <= t <= 59`.

avras replied:
I will add a comment here as well

adr1anh reviewed (commented):
Very nice work! The code is clear and makes good use of lower-level primitives which abstracts away the risks of writing circuits. All allocations use proper constructors, and variables are always derived from previous inputs. 

I left some comments suggesting the use of arrays instead of `Vec` for lists whose size are known at compile-time, and also suggestions for comments which could clarify the rounds. Let me know what you think.

adr1anh commented:
Thanks for replying to the comments, happy to review the updates!

avras commented:
Hey @adr1anh, please check if I have addressed all your suggestions. I renamed some variables to better match the spec. 

No nitpick is too small. Send them all.

adr1anh reviewed (approved):
Thanks for these changes. It looks good to me.

adr1anh reviewed (approved):
These changes look good to me, thanks for the work!
